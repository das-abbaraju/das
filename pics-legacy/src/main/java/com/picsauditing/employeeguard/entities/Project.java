package com.picsauditing.employeeguard.entities;

import org.hibernate.annotations.Where;

import javax.persistence.*;
import java.util.Date;
import java.util.List;

@Entity
@Table(name = "project")
public class Project implements BaseEntity {
	@Id
	private int id;
	private int accountID;
	private String name;
	private String location;
	private Date startDate;
	private Date endDate;

	@OneToMany(mappedBy = "project", cascade = CascadeType.ALL, orphanRemoval = true)
	@Where(clause = "deletedDate IS NULL AND deletedBy = 0")
	private List<ProjectSkill> skills;

	@OneToMany(mappedBy = "project", cascade = CascadeType.ALL, orphanRemoval = true)
	@Where(clause = "deletedDate IS NULL AND deletedBy = 0")
	private List<ProjectGroup> groups;

	@OneToMany(mappedBy = "project", cascade = CascadeType.ALL, orphanRemoval = true)
	@Where(clause = "deletedDate IS NULL AND deletedBy = 0")
	private List<ProjectCompany> companies;

	private int createdBy;
	private Date createdDate;
	private int updatedBy;
	private Date updatedDate;
	private int deletedBy;
	private Date deletedDate;

	@Override
	public int getId() {
		return id;
	}

	@Override
	public void setId(int id) {
		this.id = id;
	}

	public int getAccountID() {
		return accountID;
	}

	public void setAccountID(int accountID) {
		this.accountID = accountID;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public String getLocation() {
		return location;
	}

	public void setLocation(String location) {
		this.location = location;
	}

	public Date getStartDate() {
		return startDate;
	}

	public void setStartDate(Date startDate) {
		this.startDate = startDate;
	}

	public Date getEndDate() {
		return endDate;
	}

	public void setEndDate(Date endDate) {
		this.endDate = endDate;
	}

	public List<ProjectSkill> getSkills() {
		return skills;
	}

	public void setSkills(List<ProjectSkill> skills) {
		this.skills = skills;
	}

	public List<ProjectGroup> getGroups() {
		return groups;
	}

	public void setGroups(List<ProjectGroup> groups) {
		this.groups = groups;
	}

	public List<ProjectCompany> getCompanies() {
		return companies;
	}

	public void setCompanies(List<ProjectCompany> companies) {
		this.companies = companies;
	}

	@Override
	public int getCreatedBy() {
		return createdBy;
	}

	@Override
	public void setCreatedBy(int createdBy) {
		this.createdBy = createdBy;
	}

	@Override
	public int getUpdatedBy() {
		return updatedBy;
	}

	@Override
	public void setUpdatedBy(int updatedBy) {
		this.updatedBy = updatedBy;
	}

	@Override
	public int getDeletedBy() {
		return deletedBy;
	}

	@Override
	public void setDeletedBy(int deletedBy) {
		this.deletedBy = deletedBy;
	}

	@Override
	public Date getCreatedDate() {
		return createdDate;
	}

	@Override
	public void setCreatedDate(Date createdDate) {
		this.createdDate = createdDate;
	}

	@Override
	public Date getUpdatedDate() {
		return updatedDate;
	}

	@Override
	public void setUpdatedDate(Date updatedDate) {
		this.updatedDate = updatedDate;
	}

	@Override
	public Date getDeletedDate() {
		return deletedDate;
	}

	@Override
	public void setDeletedDate(Date deletedDate) {
		this.deletedDate = deletedDate;
	}
}
