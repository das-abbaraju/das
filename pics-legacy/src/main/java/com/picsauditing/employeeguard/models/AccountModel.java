package com.picsauditing.employeeguard.models;

public final class AccountModel implements Comparable<AccountModel> {

    private final int id;
    private final String name;
    private final AccountType accountType;

    public AccountModel(Builder builder) {
        this.id = builder.id;
        this.name = builder.name;
        this.accountType = builder.accountType;
    }

    public int getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public AccountType getAccountType() {
        return accountType;
    }

    public boolean isCorporate() {
        return accountType == AccountType.CORPORATE;
    }

    @Override
    public int compareTo(AccountModel that) {
        return this.name.compareToIgnoreCase(that.name);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        AccountModel that = (AccountModel) o;

        if (id != that.id) return false;
        if (accountType != that.accountType) return false;
        if (name != null ? !name.equals(that.name) : that.name != null) return false;

        return true;
    }

    @Override
    public int hashCode() {
        int result = id;
        result = 31 * result + (name != null ? name.hashCode() : 0);
        result = 31 * result + (accountType != null ? accountType.hashCode() : 0);
        return result;
    }

    public static class Builder {

        private int id;
        private String name;
        private AccountType accountType;

        public Builder id(int id) {
            this.id = id;
            return this;
        }

        public Builder name(String name) {
            this.name = name;
            return this;
        }

        public Builder accountType(AccountType accountType) {
            this.accountType = accountType;
            return this;
        }

        public AccountModel build() {
            return new AccountModel(this);
        }
    }
}
