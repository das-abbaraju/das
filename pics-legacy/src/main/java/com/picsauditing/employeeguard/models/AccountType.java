package com.picsauditing.employeeguard.models;

public enum AccountType {

    ADMIN_ACCOUNT, ASSESSMENT, CONTRACTOR, CORPORATE, OPERATOR, EMPLOYEE

}
