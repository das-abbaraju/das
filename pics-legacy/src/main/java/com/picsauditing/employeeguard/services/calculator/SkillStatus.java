package com.picsauditing.employeeguard.services.calculator;

public enum SkillStatus {

    // TODO: Rename expiring to something more meaningful that the Skill is about to expire

    // Keep in order of serverity, from highest severity to lowest
    Expired("expired"),
    Expiring("expiring"),
    Pending("pending"),
    Complete("complete");

    private String displayValue;

    private SkillStatus(String displayValue) {
        this.displayValue = displayValue;
    }

    public boolean isExpired() {
        return this == Expired;
    }

    public boolean isExpiring() {
        return this == Expiring;
    }

    public boolean isPending() {
        return this == Pending;
    }

    public boolean isComplete() {
        return this == Complete;
    }

    public String getDisplayValue() {
        return displayValue;
    }
}
