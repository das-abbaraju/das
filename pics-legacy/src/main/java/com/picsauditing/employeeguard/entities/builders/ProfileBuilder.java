package com.picsauditing.employeeguard.entities.builders;

import com.picsauditing.employeeguard.entities.Profile;
import com.picsauditing.employeeguard.entities.ProfileDocument;

import java.util.List;

public class ProfileBuilder {

	private final Profile profile;

	public ProfileBuilder() {
		this.profile = new Profile();
	}

	public ProfileBuilder id(final int id) {
		profile.setId(id);
		return this;
	}

	public ProfileBuilder documents(final List<ProfileDocument> documents) {
		profile.setDocuments(documents);
		return this;
	}

	public Profile build() {
		return profile;
	}
}
