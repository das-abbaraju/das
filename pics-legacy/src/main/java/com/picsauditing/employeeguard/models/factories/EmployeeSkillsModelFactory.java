package com.picsauditing.employeeguard.models.factories;

import com.picsauditing.employeeguard.services.status.SkillStatus;

import java.util.Set;

public class EmployeeSkillsModelFactory {

	public EmployeeSkillsModel create(final SkillStatus overallStatus,
									  final Set<CompanyStatusModelFactory.CompanyStatusModel> companyStatusModels) {
		EmployeeSkillsModel employeeSkillsModel = new EmployeeSkillsModel();

		employeeSkillsModel.setStatus(overallStatus);
		employeeSkillsModel.setSites(companyStatusModels);

		return employeeSkillsModel;
	}

	public class EmployeeSkillsModel {

		private SkillStatus status;
		private Set<CompanyStatusModelFactory.CompanyStatusModel> sites;

		public SkillStatus getStatus() {
			return status;
		}

		public void setStatus(SkillStatus status) {
			this.status = status;
		}

		public Set<CompanyStatusModelFactory.CompanyStatusModel> getSites() {
			return sites;
		}

		public void setSites(Set<CompanyStatusModelFactory.CompanyStatusModel> sites) {
			this.sites = sites;
		}
	}
}
