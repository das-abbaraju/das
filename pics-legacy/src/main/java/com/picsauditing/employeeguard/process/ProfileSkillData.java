package com.picsauditing.employeeguard.process;

import com.picsauditing.employeeguard.entities.AccountSkill;
import com.picsauditing.employeeguard.entities.Project;
import com.picsauditing.employeeguard.models.AccountModel;
import com.picsauditing.employeeguard.services.calculator.SkillStatus;

import java.util.Map;
import java.util.Set;

public class ProfileSkillData {

	// All Corporate/Site Required Skills + SiteAssignment Role Skills (for Roles not in Projects)
	// All Contractor Group Skills + Skills Required for all Employees
	private Map<AccountModel, Set<AccountSkill>> allRequiredSkills;

	private Map<AccountModel, Set<Project>> siteProjects;
 	private Map<Project, SkillStatus> projectStatuses;
	private Map<AccountModel, SkillStatus> siteStatuses;
	private Map<Project, Set<AccountSkill>> allProjectSkills;
	private Map<AccountSkill, SkillStatus> skillStatusMap;
	private SkillStatus overallStatus;
	private Map<Integer, AccountModel> contractorAccounts;
	private Map<Integer, AccountModel> siteAccounts;
	private Map<Integer, AccountModel> allAccounts;
	private Map<AccountModel, Set<AccountModel>> parentSites;

	public Map<AccountModel, Set<AccountSkill>> getAllRequiredSkills() {

		return allRequiredSkills;
	}

	public void setAllRequiredSkills(Map<AccountModel, Set<AccountSkill>> allRequiredSkills) {
		this.allRequiredSkills = allRequiredSkills;
	}

	public Map<AccountModel, Set<Project>> getSiteProjects() {
		return siteProjects;
	}

	public void setSiteProjects(Map<AccountModel, Set<Project>> siteProjects) {
		this.siteProjects = siteProjects;
	}

	public Map<Project, SkillStatus> getProjectStatuses() {
		return projectStatuses;
	}

	public void setProjectStatuses(Map<Project, SkillStatus> projectStatuses) {
		this.projectStatuses = projectStatuses;
	}

	public Map<AccountModel, SkillStatus> getSiteStatuses() {
		return siteStatuses;
	}

	public void setSiteStatuses(Map<AccountModel, SkillStatus> siteStatuses) {
		this.siteStatuses = siteStatuses;
	}

	public Map<Project, Set<AccountSkill>> getAllProjectSkills() {
		return allProjectSkills;
	}

	public void setAllProjectSkills(Map<Project, Set<AccountSkill>> allProjectSkills) {
		this.allProjectSkills = allProjectSkills;
	}

	public Map<AccountSkill, SkillStatus> getSkillStatusMap() {
		return skillStatusMap;
	}

	public void setSkillStatusMap(Map<AccountSkill, SkillStatus> skillStatusMap) {
		this.skillStatusMap = skillStatusMap;
	}

	public SkillStatus getOverallStatus() {
		return overallStatus;
	}

	public void setOverallStatus(SkillStatus overallStatus) {
		this.overallStatus = overallStatus;
	}

	public Map<Integer, AccountModel> getContractorAccounts() {
		return contractorAccounts;
	}

	public void setContractorAccounts(Map<Integer, AccountModel> contractorAccounts) {
		this.contractorAccounts = contractorAccounts;
	}

	public Map<Integer, AccountModel> getSiteAccounts() {
		return siteAccounts;
	}

	public void setSiteAccounts(Map<Integer, AccountModel> siteAccounts) {
		this.siteAccounts = siteAccounts;
	}

	public Map<Integer, AccountModel> getAllAccounts() {
		return allAccounts;
	}

	public void setAllAccounts(Map<Integer, AccountModel> allAccounts) {
		this.allAccounts = allAccounts;
	}

	public Map<AccountModel, Set<AccountModel>> getParentSites() {
		return parentSites;
	}

	public void setParentSites(Map<AccountModel, Set<AccountModel>> parentSites) {
		this.parentSites = parentSites;
	}
}
