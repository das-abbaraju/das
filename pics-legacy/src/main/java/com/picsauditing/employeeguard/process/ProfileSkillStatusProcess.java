package com.picsauditing.employeeguard.process;

import com.picsauditing.employeeguard.entities.*;
import com.picsauditing.employeeguard.models.AccountModel;
import com.picsauditing.employeeguard.services.AccountService;
import com.picsauditing.employeeguard.services.StatusCalculatorService;
import com.picsauditing.employeeguard.services.calculator.SkillStatus;
import com.picsauditing.employeeguard.services.entity.ProfileEntityService;
import com.picsauditing.employeeguard.services.entity.RoleEntityService;
import com.picsauditing.employeeguard.services.entity.SkillEntityService;
import com.picsauditing.employeeguard.util.PicsCollectionUtil;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class ProfileSkillStatusProcess {

	@Autowired
	private AccountService accountService;
	@Autowired
	private ProfileEntityService profileEntityService;
	@Autowired
	private RoleEntityService roleEntityService;
	@Autowired
	private SkillEntityService skillEntityService;
	@Autowired
	private StatusCalculatorService statusCalculatorService;

	@Autowired
	private ProcessHelper processHelper;

	// Required Skills contains all the group skills, Site and Corporate Required Skills, Skills from Roles
	// the person is assigned to at the Site

	// Only Project required skills and skills for roles under a project show up listed under the project

	public ProfileSkillData buildProfileSkillData(final Profile profile) {
		Set<Project> projects = findProjects(profile);
		Set<Role> roles = findRoles(profile);
		Map<Project, Set<Role>> projectRoles = processHelper.getProjectRoles(projects);
		Map<Role, Set<AccountSkill>> roleSkills = processHelper.getRoleSkills(roles);
		Map<Project, Set<AccountSkill>> projectRequiredSkills = processHelper.getProjectRequiredSkills(projects);

		ProfileSkillData profileSkillData = new ProfileSkillData();

		profileSkillData.setAllProjectSkills(processHelper.allProjectSkills(projects,
				projectRequiredSkills, projectRoles, roleSkills));

		profileSkillData = addAccountInformation(profileSkillData, profile);
		profileSkillData = addAccountRequiredSkills(profileSkillData, profile, projects, projectRoles, roleSkills);
		profileSkillData = addSiteProjects(profileSkillData, projects);
		profileSkillData = addProjectStatus(profileSkillData, projects, profile);
		profileSkillData = addAccountStatus(profileSkillData, profile);
		profileSkillData = addAllSkillStatuses(profileSkillData, profile);
		profileSkillData = addOverallSkillStatus(profileSkillData);

		return profileSkillData;
	}

	private ProfileSkillData addOverallSkillStatus(final ProfileSkillData profileSkillData) {
		profileSkillData.setOverallStatus(statusCalculatorService
				.calculateOverallStatus(profileSkillData.getSkillStatusMap().values()));

		return profileSkillData;
	}

	private ProfileSkillData addAllSkillStatuses(final ProfileSkillData profileSkillData, final Profile profile) {
		Map<AccountSkill, SkillStatus> allSkillStatuses = statusCalculatorService
				.getSkillStatuses(profile.getEmployees().get(0), aggregateAllSkills(profileSkillData));

		profileSkillData.setSkillStatusMap(allSkillStatuses);

		return profileSkillData;
	}

	private Set<AccountSkill> aggregateAllSkills(final ProfileSkillData profileSkillData) {
		Set<AccountSkill> allSkills = new HashSet<>();

		allSkills.addAll(PicsCollectionUtil.mergeCollectionOfCollections(profileSkillData.getAllProjectSkills().values()));
		allSkills.addAll(PicsCollectionUtil.mergeCollectionOfCollections(profileSkillData.getAllRequiredSkills().values()));

		return allSkills;
	}

	private ProfileSkillData addAccountStatus(final ProfileSkillData profileSkillData,
											  final Profile profile) {
		Map<AccountModel, Set<AccountSkill>> allSiteSkills =
				processHelper.allSkillsForAllSite(profileSkillData.getSiteProjects(),
						profileSkillData.getAllProjectSkills(), profileSkillData.getAllRequiredSkills());

		profileSkillData.setSiteStatuses(statusCalculatorService.getSkillStatusPerEntity(profile.getEmployees().get(0), allSiteSkills));

		return profileSkillData;
	}

	private ProfileSkillData addProjectStatus(final ProfileSkillData profileSkillData,
											  final Set<Project> projects,
											  final Profile profile) {


		Map<Project, Set<AccountSkill>> allSkillsRequiredByProject =
				processHelper.aggregateAllSkillsForProjects(projects, profileSkillData.getSiteAccounts(),
						profileSkillData.getAllProjectSkills(), profileSkillData.getAllRequiredSkills());

		profileSkillData.setProjectStatuses(statusCalculatorService
				.getSkillStatusPerEntity(profile.getEmployees().get(0), allSkillsRequiredByProject));

		return profileSkillData;
	}

	private ProfileSkillData addSiteProjects(final ProfileSkillData profileSkillData,
											 final Set<Project> projects) {
		profileSkillData.setSiteProjects(processHelper.getSiteProjects(profileSkillData.getSiteAccounts(),
				processHelper.getProjectsBySite(projects)));

		return profileSkillData;
	}

	private ProfileSkillData addAccountRequiredSkills(final ProfileSkillData profileSkillData,
													  final Profile profile,
													  final Set<Project> projects,
													  final Map<Project, Set<Role>> projectRoles,
													  final Map<Role, Set<AccountSkill>> roleSkills) {
		// Add Contractor Required Skills
		Map<AccountModel, Set<AccountSkill>> allContractorSkills = buildContractorSkillsMap(profileSkillData, profile);

		// Add Corp/Site Required Skills + Site Assignment Roles Skills not part of projects
		Map<AccountModel, Set<AccountSkill>> allSiteRequiredSkills =
				buildSiteRequiredSkillsMap(profileSkillData, profile, projects, projectRoles, roleSkills);

		profileSkillData.setAllRequiredSkills(PicsCollectionUtil.mergeMapOfSets(allContractorSkills,
				allSiteRequiredSkills));

		return profileSkillData;  //To change body of created methods use File | Settings | File Templates.
	}

	private Map<AccountModel, Set<AccountSkill>> buildSiteRequiredSkillsMap(final ProfileSkillData profileSkillData,
																			final Profile profile,
																			final Set<Project> projects,
																			final Map<Project, Set<Role>> projectRoles,
																			final Map<Role, Set<AccountSkill>> roleSkills) {
		Map<AccountModel, Set<AccountSkill>> siteRequiredSkills = findAllRequiredSkillsForEachSite(profileSkillData.getParentSites());
		Map<AccountModel, Set<AccountSkill>> roleSkillsNotForProjects =
				findAllRoleSkillsNotForProjects(profileSkillData.getSiteAccounts(), projects, projectRoles,
						roleSkills, profile);

		return PicsCollectionUtil.mergeMapOfSets(siteRequiredSkills, roleSkillsNotForProjects);
	}

	private Map<AccountModel, Set<AccountSkill>> findAllRequiredSkillsForEachSite(final Map<AccountModel, Set<AccountModel>> parentSitesMap) {
		Map<AccountModel, Set<AccountSkill>> siteRequiredSkillsMap = new HashMap<>();
		for (AccountModel site : parentSitesMap.keySet()) {
			Set<AccountSkill> requiredSkills = skillEntityService.getSiteAndCorporateRequiredSkills(site.getId(),
					PicsCollectionUtil.getIdsFromCollection(parentSitesMap.get(site),

							new PicsCollectionUtil.Identitifable<AccountModel, Integer>() {

								@Override
								public Integer getId(AccountModel accountModel) {
									return accountModel.getId();
								}
							}));

			siteRequiredSkillsMap.put(site, requiredSkills);
		}

		return siteRequiredSkillsMap;
	}

	private Map<AccountModel, Set<AccountSkill>> findAllRoleSkillsNotForProjects(final Map<Integer, AccountModel> siteAccounts,
																				 final Set<Project> projects,
																				 final Map<Project, Set<Role>> projectRoles,
																				 final Map<Role, Set<AccountSkill>> roleSkills,
																				 final Profile profile) {
		Map<Integer, Set<Project>> siteProjects = processHelper.getProjectsBySite(projects);
		Map<Integer, Set<Role>> siteRoles = processHelper.getRolesBySite(profile.getEmployees());
		Map<Integer, Set<Role>> siteRolesNotInProjects = processHelper
				.siteRolesNotInProjects(siteProjects, siteRoles, projectRoles);

		Map<AccountModel, Set<AccountSkill>> roleSkillsNotForProjects = new HashMap<>();
		for (Integer siteId : siteRolesNotInProjects.keySet()) {
			Set<AccountSkill> skillsForRole = roleSkills.get(siteRolesNotInProjects.get(siteId));
			if (skillsForRole == null) {
				skillsForRole = new HashSet<>();
			}

			roleSkillsNotForProjects.put(siteAccounts.get(siteId), skillsForRole);
		}

		return roleSkillsNotForProjects;
	}

	private Map<AccountModel, Set<AccountSkill>> buildContractorSkillsMap(final ProfileSkillData profileSkillData,
																		  final Profile profile) {

		Map<Integer, Set<AccountSkill>> contractorSkillsForGroups = getGroupSkillsByContractorAccountId(profile);

		Map<Integer, AccountModel> contractorAccounts = profileSkillData.getContractorAccounts();
		Map<Integer, Set<AccountSkill>> contractorRequiredSkills =
				processHelper.contractorRequiredSkills(contractorAccounts.keySet());

		Map<Integer, Set<AccountSkill>> allContractorSkillsByAccountId = PicsCollectionUtil
				.mergeMapOfSets(contractorSkillsForGroups, contractorRequiredSkills);

		Map<AccountModel, Set<AccountSkill>> contractorRequiredSkill = new HashMap<>();
		for (Integer accountId : contractorAccounts.keySet()) {
			Set<AccountSkill> allContractorSkills = allContractorSkillsByAccountId.get(accountId);
			if (allContractorSkills == null) {
				allContractorSkills = new HashSet<>();
			}

			contractorRequiredSkill.put(contractorAccounts.get(accountId), allContractorSkills);
		}

		return contractorRequiredSkill;
	}

	private Map<Integer, Set<AccountSkill>> getGroupSkillsByContractorAccountId(final Profile profile) {
		Map<Group, Set<AccountSkill>> groupSkills = processHelper.getGroupSkills(profile);
		return PicsCollectionUtil.transformMap(groupSkills,

				new PicsCollectionUtil.KeyTransformable<Group, Integer>() {

					@Override
					public Integer getNewKey(Group group) {
						return group.getAccountId();
					}
				});
	}

	private ProfileSkillData addAccountInformation(final ProfileSkillData profileSkillData, final Profile profile) {
		Set<Integer> contractorIds = allContractors(profile);
		Map<Integer, AccountModel> contractorAccounts = accountService.getIdToAccountModelMap(contractorIds);

		Set<Integer> siteIds = allSites(profile);
		Map<Integer, AccountModel> siteAccounts = accountService.getIdToAccountModelMap(siteIds);

		profileSkillData.setContractorAccounts(contractorAccounts);
		profileSkillData.setSiteAccounts(siteAccounts);
		profileSkillData.setParentSites(accountService.getSiteParentAccounts(siteIds));
		profileSkillData.setAllAccounts(PicsCollectionUtil.mergeMaps(contractorAccounts, siteAccounts));

		return profileSkillData;
	}

	private Set<Integer> allContractors(final Profile profile) {
		return PicsCollectionUtil.extractPropertyToSet(profileEntityService.getEmployeesForProfile(profile),

				new PicsCollectionUtil.PropertyExtractor<Employee, Integer>() {
					@Override
					public Integer getProperty(Employee employee) {
						return employee.getAccountId();
					}
				});
	}

	private Set<Integer> allSites(final Profile profile) {
		return profileEntityService.getSiteAssignments(profile);
	}

	private Map<Integer, Set<AccountSkill>> getAllRequiredSkills(final Set<Integer> siteAndCorporateIds) {
		return skillEntityService.getSiteRequiredSkills(siteAndCorporateIds);
	}

	private Set<Project> findProjects(final Profile profile) {
		return profileEntityService.findProjectsForProfile(profile);
	}

	private Set<Role> findRoles(final Profile profile) {
		return roleEntityService.getRolesForProfile(profile);
	}

	// TODO: Move these methods to another class
}
