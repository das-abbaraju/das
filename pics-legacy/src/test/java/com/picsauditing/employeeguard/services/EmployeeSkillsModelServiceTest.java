package com.picsauditing.employeeguard.services;

import com.google.gson.Gson;
import com.picsauditing.employeeguard.entities.AccountSkill;
import com.picsauditing.employeeguard.entities.Profile;
import com.picsauditing.employeeguard.entities.Project;
import com.picsauditing.employeeguard.entities.builders.AccountSkillBuilder;
import com.picsauditing.employeeguard.entities.builders.ProfileBuilder;
import com.picsauditing.employeeguard.entities.builders.ProjectBuilder;
import com.picsauditing.employeeguard.models.AccountModel;
import com.picsauditing.employeeguard.models.AccountType;
import com.picsauditing.employeeguard.models.factories.EmployeeSkillsModelFactory;
import com.picsauditing.employeeguard.process.ProfileSkillData;
import com.picsauditing.employeeguard.process.ProfileSkillStatusProcess;
import com.picsauditing.employeeguard.services.calculator.SkillStatus;
import org.approvaltests.Approvals;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.UseReporter;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.powermock.reflect.Whitebox;

import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Set;

import static org.mockito.Mockito.when;

@UseReporter(DiffReporter.class)
public class EmployeeSkillsModelServiceTest {

	public static final int CORPORATE_ID = 15;
	public static final AccountModel CORPORATE_ACCOUNT_MODEL = new AccountModel.Builder().id(CORPORATE_ID).name("Corp Account").accountType(AccountType.CORPORATE).build();
	public static final int SITE_ID = 67;
	public static final AccountModel SITE_ACCOUNT_MODEL = new AccountModel.Builder().id(SITE_ID).name("Site Account").accountType(AccountType.OPERATOR).build();
	public static final int CONTRACTOR_ID = 120;
	public static final AccountModel CONTRACTOR_ACCOUNT_MODEL = new AccountModel.Builder().id(CONTRACTOR_ID).name("Contractor Account").accountType(AccountType.CONTRACTOR).build();

	public static final Project TEST_PROJECT = new ProjectBuilder().accountId(SITE_ID).name("Test Project").build();
	public static final AccountSkill PROJECT_REQUIRED_SKILL_1 = new AccountSkillBuilder().accountId(CORPORATE_ID).name("Project Skill 1").build();
	public static final AccountSkill PROJECT_REQUIRED_SKILL_2 = new AccountSkillBuilder().accountId(CORPORATE_ID).name("Project Skill 2").build();
	public static final AccountSkill CORPORATE_REQUIRED_SKILL = new AccountSkillBuilder().accountId(CORPORATE_ID).name("Corp Required Skill").build();
	public static final AccountSkill SITE_REQUIRED_SKILL = new AccountSkillBuilder().accountId(CORPORATE_ID).name("Site Required Skill").build();
	public static final AccountSkill CONTRACTOR_REQUIRED_SKILL = new AccountSkillBuilder().accountId(CONTRACTOR_ID).name("Contractor RequiredSkill").build();

	// Class under test
	private EmployeeSkillsModelService employeeSkillsModelService;

	@Mock
	private ProfileSkillStatusProcess profileSkillStatusProcess;

	@Before
	public void setUp() throws Exception {
		MockitoAnnotations.initMocks(this);

		employeeSkillsModelService = new EmployeeSkillsModelService();

		Whitebox.setInternalState(employeeSkillsModelService, "profileSkillStatusProcess", profileSkillStatusProcess);
	}

	@Test
	public void testBuildEmployeeSkillsModel() throws Exception {
		Profile fakeProfile = new ProfileBuilder().id(890).build();
		when(profileSkillStatusProcess.buildProfileSkillData(fakeProfile)).thenReturn(buildFakeProfileSkillData());

		EmployeeSkillsModelFactory.EmployeeSkillsModel result = employeeSkillsModelService
				.buildEmployeeSkillsModel(fakeProfile);

		verifyTestBuildEmployeeSkillsModel(result);
	}

	private void verifyTestBuildEmployeeSkillsModel(EmployeeSkillsModelFactory.EmployeeSkillsModel result) throws Exception {
		Approvals.verify(new Gson().toJson(result));
	}

	private ProfileSkillData buildFakeProfileSkillData() {
		ProfileSkillData profileSkillData = new ProfileSkillData();

		profileSkillData = addProjects(profileSkillData);
		profileSkillData = addAllProjectSkills(profileSkillData);
		profileSkillData = addSkillStatusMap(profileSkillData);
		profileSkillData = addSiteAccounts(profileSkillData);
		profileSkillData = addProjectStatuses(profileSkillData);
		profileSkillData = addAllProjectSkills(profileSkillData);
		profileSkillData = addAllRequiredSkills(profileSkillData);
		profileSkillData = addOverallStatus(profileSkillData);
		profileSkillData = addSiteStatuses(profileSkillData);

		return profileSkillData;
	}

	private ProfileSkillData addProjects(final ProfileSkillData profileSkillData) {
		profileSkillData.setProjects(new HashSet<>(Arrays.asList(TEST_PROJECT)));

		return profileSkillData;
	}

	private ProfileSkillData addAllProjectSkills(final ProfileSkillData profileSkillData) {
		profileSkillData.setAllProjectSkills(new HashMap<Project, Set<AccountSkill>>() {{

			put(TEST_PROJECT, new HashSet<>(Arrays.asList(PROJECT_REQUIRED_SKILL_1, PROJECT_REQUIRED_SKILL_2)));

		}});

		return profileSkillData;
	}

	private ProfileSkillData addSkillStatusMap(final ProfileSkillData profileSkillData) {
		profileSkillData.setSkillStatusMap(new HashMap<AccountSkill, SkillStatus>() {{

			put(PROJECT_REQUIRED_SKILL_1, SkillStatus.Completed);
			put(PROJECT_REQUIRED_SKILL_2, SkillStatus.Completed);
			put(CORPORATE_REQUIRED_SKILL, SkillStatus.Expiring);
			put(SITE_REQUIRED_SKILL, SkillStatus.Completed);
			put(CONTRACTOR_REQUIRED_SKILL, SkillStatus.Completed);

		}});

		return profileSkillData;
	}

	private ProfileSkillData addSiteAccounts(final ProfileSkillData profileSkillData) {
		profileSkillData.setSiteAccounts(new HashMap<Integer, AccountModel>() {{

			put(CORPORATE_ID, CORPORATE_ACCOUNT_MODEL);
			put(SITE_ID, SITE_ACCOUNT_MODEL);

		}});

		return profileSkillData;
	}

	private ProfileSkillData addProjectStatuses(final ProfileSkillData profileSkillData) {
		profileSkillData.setProjectStatuses(new HashMap<Project, SkillStatus>() {{

			put(TEST_PROJECT, SkillStatus.Completed);

		}});

		return profileSkillData;
	}

	private ProfileSkillData addAllRequiredSkills(final ProfileSkillData profileSkillData) {
		profileSkillData.setAllRequiredSkills(new HashMap<AccountModel, Set<AccountSkill>>() {{

			put(CORPORATE_ACCOUNT_MODEL, new HashSet<>(Arrays.asList(CORPORATE_REQUIRED_SKILL)));
			put(SITE_ACCOUNT_MODEL, new HashSet<>(Arrays.asList(SITE_REQUIRED_SKILL, CORPORATE_REQUIRED_SKILL)));
			put(CONTRACTOR_ACCOUNT_MODEL, new HashSet<>(Arrays.asList(CONTRACTOR_REQUIRED_SKILL)));

		}});

		return profileSkillData;
	}

	private ProfileSkillData addOverallStatus(final ProfileSkillData profileSkillData) {
		profileSkillData.setOverallStatus(SkillStatus.Expiring);

		return profileSkillData;
	}

	private ProfileSkillData addSiteStatuses(final ProfileSkillData profileSkillData) {
		profileSkillData.setSiteStatuses(new HashMap<AccountModel, SkillStatus>() {{

			put(CORPORATE_ACCOUNT_MODEL, SkillStatus.Expiring);
			put(SITE_ACCOUNT_MODEL, SkillStatus.Completed);
			put(CONTRACTOR_ACCOUNT_MODEL, SkillStatus.Completed);

		}});

		return profileSkillData;
	}
}
