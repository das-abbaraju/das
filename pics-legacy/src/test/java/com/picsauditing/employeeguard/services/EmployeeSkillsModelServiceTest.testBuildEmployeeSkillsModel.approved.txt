{"status":"Expiring","sites":[{"id":120,"name":"Contractor Account","status":"Completed","skills":[{"id":0,"name":"Contractor RequiredSkill","status":"Completed"}]},{"id":15,"name":"Corp Account","status":"Expiring","skills":[{"id":0,"name":"Corp Required Skill","status":"Expiring"}]},{"id":67,"name":"Site Account","status":"Completed","skills":[{"id":0,"name":"Corp Required Skill","status":"Expiring"},{"id":0,"name":"Site Required Skill","status":"Completed"}],"projects":[{"id":0,"name":"Test Project","status":"Completed","skills":[{"id":0,"name":"Corp Required Skill","status":"Expiring"},{"id":0,"name":"Project Skill 1","status":"Completed"},{"id":0,"name":"Project Skill 2","status":"Completed"},{"id":0,"name":"Site Required Skill","status":"Completed"}]}]}]}